swift/ql/lib/codeql/swift/frameworks/StandardLibrary/WebView.qll

```diff
@@ -33,6 +33,58 @@ private class WKScriptMessageBodyInheritsTaint extends TaintInheritingContent,
   }
 }
 
+/**
+ * A type or extension delcaration that adopts the protocol `WKNavigationDelegate`.
+ */
+private class AdoptsWkNavigationDelegate extends Decl {
+  AdoptsWkNavigationDelegate() {
+    exists(ProtocolDecl delegate |
+      this.(ExtensionDecl).getAProtocol().getABaseTypeDecl*() = delegate or
+      this.(NominalTypeDecl).getABaseTypeDecl*() = delegate
+    |
+      delegate.getName() = "WKNavigationDelegate"
+    )
+  }
+}
+
+/**
+ * A source for the `decidePolicyFor` parameter of the `webView` method of a type adopting `WKNavigationDelegate`.
+ */
+private class WKNavigationDelegateSource extends RemoteFlowSource {
+  WKNavigationDelegateSource() {
+    exists(ParamDecl p, FuncDecl f, AdoptsWkNavigationDelegate t |
+      f.getEnclosingDecl() = t and
+      f.getName() =
+        [
+          "webView(_:decidePolicyFor:preferences:decisionHandler:)",
+          "webView(_:decidePolicyFor:decisionHandler:)"
+        ] and
+      p.getDeclaringFunction() = f and
+      p.getIndex() = 1 and
+      this.(DataFlow::ParameterNode).getParameter() = p
+    )
+  }
+
+  override string getSourceType() { result = "Navigation action of a WebView" }
+}
+
+/**
+ * A taint step implying that, if a `WKNavigationAction` is tainted, its `request` field is also tainted.
+ */
+private class WKNavigationActionTaintStep extends AdditionalTaintStep {
+  override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
+    exists(MemberRefExpr e, Expr self, VarDecl member |
+      self.getType().getName() = "WKNavigationAction" and
+      member.getName() = "request"
+    |
+      e.getBase() = self and
+      e.getMember() = member and
+      n1.asExpr() = self and
+      n2.asExpr() = e
```

Comment:
What's the reason for choosing this design rather than using `TaintInheritingContent`?  Should we prefer one or the other approach in general?

Reply:
They are mostly equivalent, but I slightly prefer `TaintInheritingContent` because it's shorter.

The reason I instead used a step here is because, in the project I was testing this, I found a case where the `.request` access pointed to a `request` `ConcreteVarDecl` (not `FieldDecl`) that appeared to not have a declaring/enclosing declaration. Which made `TaintInheritingContent` not work, but the step did.

Now that I wrote this, that's probably a bad reason to use a taint step instead of a `TaintInheritingContent`, because I'm pretty sure the problem has something to do with the extractor, so once that's fixed, the `TaintInheritingContent` should work. The problem is that this doesn't seem to be easily reproducible in tests.

What do you think? Is it worth it keeping it as a step until that's fixed? It most probably will mean that it'll be a step forever though...

Reply:
If it works better right now, lets keep it as it is.  You can either insert a TODO comment or create an issue if you'd like to see it fixed in future ... though if there isn't a test demonstrating the issue, we might well lack the courage to change it.

Reply:
... ah, I see you've created an issue for this already.  Then I'm happy and I think we should merge.